azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -107,12 +99,441 @@ def mock_execute_op(self, op):
+        first_connect = pipeline_ops_base.ConnectOperation(callback=mocker.MagicMock())
+        first_fake_op = FakeOperation(callback=mocker.MagicMock())
+        second_connect = pipeline_ops_base.ReconnectOperation(callback=mocker.MagicMock())
+        second_fake_op = FakeOperation(callback=mocker.MagicMock())
+
+        stage.run_op(first_connect)
+        stage.run_op(first_fake_op)
+        stage.run_op(second_connect)
+        stage.run_op(second_fake_op)
+
+        # at this point, ops are pended waiting for the first connect to complete.  Verify this and complete the connect.
+        assert stage.next.run_op.call_count == 1
+        assert stage.next.run_op.call_args[0][0] == first_connect
+        operation_flow.complete_op(stage=stage.next, op=first_connect)
+
+        # The connect is complete.  This passes down first_fake_op and second_connect and second_fake_op gets pended waiting i
+        # for second_connect to complete.
+        # Note: this isn't ideal.  In a perfect world, second_connect wouldn't start until first_fake_op is complete, but we
+        # dont have this logic in place yet.
+        assert stage.next.run_op.call_count == 3
+        assert stage.next.run_op.call_args_list[1][0][0] == first_fake_op
+        assert stage.next.run_op.call_args_list[2][0][0] == second_connect
+
+        # now, complete second_connect to give second_fake_op a chance to get passed down
+        operation_flow.complete_op(stage=stage.next, op=second_connect)
+        assert stage.next.run_op.call_count == 4
+        assert stage.next.run_op.call_args_list[3][0][0] == second_fake_op
+
+    @pytest.mark.parametrize(
+        "params",
+        [
+            pytest.param(
+                {
+                    "pre_connected_flag": True,
+                    "first_connect_op": pipeline_ops_base.DisconnectOperation,
+                    "mid_connect_flag": False,
+                    "second_connect_op": pipeline_ops_base.DisconnectOperation,
+                },
+                id="Disconnect followed by Disconnect",
+            ),
+            pytest.param(
+                {
+                    "pre_connected_flag": False,
+                    "first_connect_op": pipeline_ops_base.ConnectOperation,
+                    "mid_connect_flag": True,
+                    "second_connect_op": pipeline_ops_base.ConnectOperation,
+                },
+                id="Connect followed by Connect",
+            ),
+            pytest.param(
+                {
+                    "pre_connected_flag": True,
+                    "first_connect_op": pipeline_ops_base.ReconnectOperation,
+                    "mid_connect_flag": True,
+                    "second_connect_op": pipeline_ops_base.ConnectOperation,
+                },
+                id="Reconnect followed by Connect",
+            ),
+        ],
+    )
+    @pytest.mark.it(
+        "Immediately completes a second op which was waiting for a first op that succeeded"
```

Comment:
>"Immediately completes a second op which was waiting for a first op that succeeded" [](start = 8, length = 83)

Wow, this one is __really__ hard too. We have to capture the relationship between the operations, and their use of the queue

Maybe "Immediately completes a queued operation whose desired result has already been completed by a previous operation"?

There should also probably be a test cate with intermediate operations i.e. Connect->Publish->Connect, since this doesn't only apply to ops in direct succession

Reply:
Added in TFS task 5117988

---
In reply to: [312678963](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r312678963) [](ancestors = 312678963)